File: web/eXoResources/src/main/webapp/javascript/eXo/webui/UIDropDownControl.js
/**
 * Copyright (C) 2009 eXo Platform SAS.
 * 
 * This is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 * 
 * This software is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this software; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA, or see the FSF
 * site: http://www.fsf.org.
 */

eXo.webui.UIDropDownControl = {

  init : function(id, webui)
  {
    var elmt = $('#' + id);
    elmt.find('.UIDropDownTitle').on('click', this.showEvt);
    elmt.find('a.OptionItem').on('click', this.onclickEvt);
    elmt.find(".TopItemContainer").on('mousedown', function() {
    	webui.VerticalScrollManager.initScroll(this, true, 10);
    });   
    elmt.find(".BottomItemContainer").on('mousedown', function() {
    	webui.VerticalScrollManager.initScroll(this, false, 10);
    });
  },

  selectItem : function(method, id, selectedIndex) {
    if (method)
      method(id, selectedIndex);
  },

  /*
   * . minh.js.exo
   */
  /**
   * show or hide drop down control
   * 
   * @param {Object}
   *          obj document object to use as Anchor for drop down
   * @param {Object}
   *          evet event object
   */
  showEvt : function() {
    var dropDownAnchor = $(this).next("div");
    if(!dropDownAnchor)
    {
      return;
    }

    if(dropDownAnchor.css("display") == "none")
    {
      dropDownAnchor.css({"display" : "block" , "visibility" : "visible"});

      var middleCont = dropDownAnchor.find("div.MiddleItemContainer");
      var topCont = middleCont.prev("div");
      var bottomCont = middleCont.next("div");

      topCont.css("display", "block");
      bottomCont.css("display", "block");

      var visibleHeight = $(window).height() - $(middleCont[0]).offset().top - 40;
      if(middleCont[0].scrollHeight > visibleHeight)
      {
        middleCont.css("height", visibleHeight - topCont[0].offsetHeight - bottomCont[0].offsetHeight + "px");
        topCont.click(function() { return false;});
        bottomCont.click(function() { return false;});
      }
      else
      {
        topCont.css("display", "none");
        bottomCont.css("display", "none");
        middleCont.scrollTop(0);
        middleCont.css("height", "auto");
      }

      $(document).one("click", function()
      {
    	  $(document).one("click", function()
    	  {
	        dropDownAnchor.css("display", "none");
    	  });
      });
    }
    else
    {
      dropDownAnchor.css({"display" : "none" , "visibility" : "hidden"});
    }
  },
  /**
   * Hide an object
   * 
   * @param {Object,
   *          String} obj object to hide
   */
  hide : function(obj) {
    if (typeof (obj) == "string")
      obj = document.getElementById(obj);
    obj.style.display = "none";
  },
  /**
   * Use as event when user selects a item in drop down list Display content of
   * selected item and hide drop down control
   * 
   * @param {Object}
   *          obj selected object
   */
  onclickEvt : function() {
    var dropDownAnchor = $(this).parents("div.UIDropDownAnchor");
    var dropDownMiddleTitle = dropDownAnchor.prev("div.UIDropDownTitle").find("div.DropDownSelectLabel");
    dropDownMiddleTitle.html($(this).html());

    dropDownAnchor.css("display", "none");
  }
};

_module.UIDropDownControl = eXo.webui.UIDropDownControl;

File: webui/core/src/main/java/org/exoplatform/webui/core/UIDropDownControl.java
/**
 * Copyright (C) 2009 eXo Platform SAS.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

package org.exoplatform.webui.core;

import java.util.ArrayList;
import java.util.List;

import org.exoplatform.webui.config.annotation.ComponentConfig;
import org.exoplatform.webui.core.model.SelectItemOption;

/**
 * Created by The eXo Platform SARL Author : Tran The Trong trongtt@gmail.com July 12, 2007
 *
 * A drop down selector
 */
@ComponentConfig(template = "system:/groovy/webui/core/UIDropDownControl.gtmpl")
public class UIDropDownControl extends UIComponent {
    /**
     * The action to perform when an item is selected
     */
    private String action_;

    /**
     * The list of items
     */
    private List<SelectItemOption<String>> options_;

    /**
     * The index in the list of the selected item
     */
    private int selectedItemIndex_ = 0;

    public UIDropDownControl() {
        options_ = new ArrayList<SelectItemOption<String>>();
    }

    public void setValue(String value) {
        int ln = options_.size();
        for (int i = 0; i < ln; i++) {
            if (options_.get(i).getValue().equals(value)) {
                selectedItemIndex_ = i;
                return;
            }
        }
    }

    public void setValue(int i) {
        selectedItemIndex_ = i;
        return;
    }

    public void cleanItem() {
        options_.clear();
    }

    public void setOptions(List<SelectItemOption<String>> options) {
        options_ = options;
        if (options == null)
            return;
        if (options_.size() < 1)
            return;
    }

    public int getSelectedIndex() {
        if (options_.size() < 1)
            return -1;
        return selectedItemIndex_;
    }

    public String getLabel() {
        return options_.get(selectedItemIndex_).getLabel();
    }

    public String getValue() {
        return options_.get(selectedItemIndex_).getValue();
    }

    public void setAction(String act) {
        action_ = act;
    }

    public String getAction() {
        return action_;
    }

    public List<SelectItemOption<String>> getOptions() {
        return options_;
    }

    public void addItem(SelectItemOption<String> opt) {
        options_.add(opt);
    }

    public void addItem(String value) {
        options_.add(new SelectItemOption<String>(value));
    }

    public void addItem(String label, String value) {
        options_.add(new SelectItemOption<String>(label, value));
    }

    public String event(int selectedIndex) throws Exception {
        if (action_ == null) {
            return super.event(config.getEvents().get(0).getName(), options_.get(selectedIndex).getValue());
        }
        StringBuilder evt = new StringBuilder("javascript:eXo.webui.UIDropDownControl.selectItem(");
        evt.append(action_).append(",'").append(this.getId()).append("','").append(selectedIndex).append("')");
        return evt.toString();
    }
}
